### ISSUE TITLE
CUDA v12 install instructions 

### ISSUE THREAD

**Issue by MKCarter** [bjing2016/alphaflow#40]
Hi,

I have managed to get a version of `alphaflow` working with `CUDA 12`. This [link](https://openfold.readthedocs.io/en/latest/Installation.html#) was a good resource for installing a CUDA 12 compatible version of `openfold`.

To install the environment:

```
conda create -n alphaflow python=3.9
pip install numpy==1.21.2 pandas==1.5.3
pip install torch==2.5.1
pip install biopython dm-tree==0.1.6 modelcif==0.7 ml-collections==0.1.0 scipy==1.7.1 absl-py einops
pip install pytorch_lightning==2.0.4 fair-esm mdtraj==1.9.9 wandb
git clone https://github.com/aqlaboratory/openfold/tree/pl_upgrades
cd openfold
pip install -e .
cd ..
```
Then, just to check the `train.py` and `predict.py` run as expected:

```
python train.py -h
usage: train.py [-h] [--mode {esmfold,alphafold}] [--ckpt CKPT] [--restore_weights_only] [--validate] [--epochs EPOCHS] [--train_epoch_len TRAIN_EPOCH_LEN]
                [--limit_batches LIMIT_BATCHES] [--batch_size BATCH_SIZE] [--num_workers NUM_WORKERS] [--check_grad] [--accumulate_grad ACCUMULATE_GRAD]
                [--grad_clip GRAD_CLIP] [--lr LR] [--no_ema] [--train_data_dir TRAIN_DATA_DIR] [--pdb_chains PDB_CHAINS] [--train_msa_dir TRAIN_MSA_DIR]
                [--pdb_clusters PDB_CLUSTERS] [--train_cutoff TRAIN_CUTOFF] [--mmcif_dir MMCIF_DIR] [--filter_chains] [--sample_train_confs] [--val_csv VAL_CSV]
                [--val_samples VAL_SAMPLES] [--val_msa_dir VAL_MSA_DIR] [--sample_val_confs] [--num_val_confs NUM_VAL_CONFS] [--normal_validate] [--noise_prob NOISE_PROB]
                [--self_cond_prob SELF_COND_PROB] [--extra_input] [--extra_input_prob EXTRA_INPUT_PROB] [--first_as_template] [--distillation] [--distill_self_cond]
                [--print_freq PRINT_FREQ] [--val_freq VAL_FREQ] [--ckpt_freq CKPT_FREQ] [--wandb] [--run_name RUN_NAME]

optional arguments:
  -h, --help            show this help message and exit
  --mode {esmfold,alphafold}
  --ckpt CKPT
  --restore_weights_only
  --validate
  --epochs EPOCHS
  --train_epoch_len TRAIN_EPOCH_LEN
  --limit_batches LIMIT_BATCHES
  --batch_size BATCH_SIZE
  --num_workers NUM_WORKERS
  --check_grad
  --accumulate_grad ACCUMULATE_GRAD
  --grad_clip GRAD_CLIP
  --lr LR
  --no_ema
  --train_data_dir TRAIN_DATA_DIR
  --pdb_chains PDB_CHAINS
  --train_msa_dir TRAIN_MSA_DIR
  --pdb_clusters PDB_CLUSTERS
  --train_cutoff TRAIN_CUTOFF
  --mmcif_dir MMCIF_DIR
  --filter_chains
  --sample_train_confs
  --val_csv VAL_CSV
  --val_samples VAL_SAMPLES
  --val_msa_dir VAL_MSA_DIR
  --sample_val_confs
  --num_val_confs NUM_VAL_CONFS
  --normal_validate
  --noise_prob NOISE_PROB
  --self_cond_prob SELF_COND_PROB
  --extra_input
  --extra_input_prob EXTRA_INPUT_PROB
  --first_as_template
  --distillation
  --distill_self_cond
  --print_freq PRINT_FREQ
  --val_freq VAL_FREQ
  --ckpt_freq CKPT_FREQ
  --wandb
  --run_name RUN_NAME
```

```
python predict.py -h
usage: predict.py [-h] [--input_csv INPUT_CSV] [--templates_dir TEMPLATES_DIR] [--msa_dir MSA_DIR] [--mode {alphafold,esmfold}] [--samples SAMPLES] [--steps STEPS]
                  [--outpdb OUTPDB] [--weights WEIGHTS] [--ckpt CKPT] [--original_weights] [--pdb_id [PDB_ID ...]] [--subsample SUBSAMPLE] [--resample] [--tmax TMAX]
                  [--no_diffusion] [--self_cond] [--noisy_first] [--runtime_json RUNTIME_JSON] [--no_overwrite]

optional arguments:
  -h, --help            show this help message and exit
  --input_csv INPUT_CSV
  --templates_dir TEMPLATES_DIR
  --msa_dir MSA_DIR
  --mode {alphafold,esmfold}
  --samples SAMPLES
  --steps STEPS
  --outpdb OUTPDB
  --weights WEIGHTS
  --ckpt CKPT
  --original_weights
  --pdb_id [PDB_ID ...]
  --subsample SUBSAMPLE
  --resample
  --tmax TMAX
  --no_diffusion
  --self_cond
  --noisy_first
  --runtime_json RUNTIME_JSON
  --no_overwrite
```

So all seems to work as expected :) 

I hope this helps anyone trying to use this with CUDA 12 updwards.

Thanks,
Mike

**Comment by Raul-araya-secchi** [comment]
Hi,

I tried the instructions but got the following error:

` [raraya@chamaco alphaflow]$ pip install numpy==1.21.2 pandas==1.5.3
ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11
ERROR: Could not find a version that satisfies the requirement numpy==1.21.2 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23.4, 1.23.5, 1.24.0, 1.24.1, 1.24.2, 1.24.3, 1.24.4, 1.25.0, 1.25.1, 1.25.2, 1.26.0, 1.26.1, 1.26.2, 1.26.3, 1.26.4, 2.0.0, 2.0.1, 2.0.2, 2.1.0rc1, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.2.0rc1)
ERROR: No matching distribution found for numpy==1.21.2`

After that I could not continue.

**Comment by MKCarter** [comment]
Yeah, you need to activate the env first in order to run the rest...

Running this will work:

```
git clone https://github.com/bjing2016/alphaflow
cd alphaflow
conda create -n alphaflow python=3.9
conda activate alphaflow
pip install numpy==1.21.2 pandas==1.5.3
pip install torch==2.5.1
pip install biopython dm-tree==0.1.6 modelcif==0.7 ml-collections==0.1.0 scipy==1.7.1 absl-py einops
pip install pytorch_lightning==2.0.4 fair-esm mdtraj==1.9.9 wandb
git clone https://github.com/aqlaboratory/openfold/tree/pl_upgrades
cd openfold
pip install -e .
cd ..
```
Sorry that wasnt clear previously

**Comment by Raul-araya-secchi** [comment]
I think there is an error here: 
`git clone https://github.com/aqlaboratory/openfold/tree/pl_upgrades
Cloning into 'pl_upgrades'...
fatal: repository 'https://github.com/aqlaboratory/openfold/tree/pl_upgrades/' not found
`
The repo is not there.

**Comment by MKCarter** [comment]
Hey, the repo is [here](https://github.com/aqlaboratory/openfold/tree/pl_upgrades) 

so running:

`git clone https://github.com/aqlaboratory/openfold.git`

will work. 

Sorry this wasn’t clear before.

**Comment by MKCarter** [comment]
Actually you need to run:

`git clone -b pl_upgrades https://github.com/aqlaboratory/openfold.git`

Then install using pip. So, to confirm:

```
git clone https://github.com/bjing2016/alphaflow
cd alphaflow
conda create -n alphaflow python=3.9
conda activate alphaflow
pip install numpy==1.21.2 pandas==1.5.3
pip install torch==2.5.1
pip install biopython dm-tree==0.1.6 modelcif==0.7 ml-collections==0.1.0 scipy==1.7.1 absl-py einops
pip install pytorch_lightning==2.0.4 fair-esm mdtraj==1.9.9 wandb
git clone -b pl_upgrades https://github.com/aqlaboratory/openfold.git
cd openfold
pip install -e .
cd ..
```

**Comment by Raul-araya-secchi** [comment]
It's working. Thanks so much!

**Comment by MKCarter** [comment]
Fantastic!

**Comment by JoshuaEPSamuel** [comment]
> Actually you need to run:
> 
> `git clone -b pl_upgrades https://github.com/aqlaboratory/openfold.git`
> 
> Then install using pip. So, to confirm:
> 
> ```
> git clone https://github.com/bjing2016/alphaflow
> cd alphaflow
> conda create -n alphaflow python=3.9
> conda activate alphaflow
> pip install numpy==1.21.2 pandas==1.5.3
> pip install torch==2.5.1
> pip install biopython dm-tree==0.1.6 modelcif==0.7 ml-collections==0.1.0 scipy==1.7.1 absl-py einops
> pip install pytorch_lightning==2.0.4 fair-esm mdtraj==1.9.9 wandb
> git clone -b pl_upgrades https://github.com/aqlaboratory/openfold.git
> cd openfold
> pip install -e .
> cd ..
> ```

I've tried using this update today, however, I still get a CUDA version error when installing openfold:
`RuntimeError: The detected CUDA version (10.1) mismatches the version that was used to compile PyTorch (12.4). Please make sure to use the same CUDA versions.`

**Comment by MKCarter** [comment]
Yeah, this is down to your system having CUDA 10… I would recommend updating your CUDA drivers if possible.

**Comment by MKCarter** [comment]
I would follow this [tutorial](https://ubuntu.com/server/docs/nvidia-drivers-installation) if you want to update your drivers to v12

**Comment by JoshuaEPSamuel** [comment]
Thanks for the link! I might have to come back to this later

I've been able to temporarily sort the environment out by following your earlier [link](https://openfold.readthedocs.io/en/latest/Installation.html#) to install a conda Openfold env, run the Openfold setup, then install the packages you mentioned earlier:

> pip install python==3.9
pip install numpy==1.21.2 pandas==1.5.3
pip install torch==2.5.1
pip install biopython dm-tree==0.1.6 modelcif==0.7 ml-collections==0.1.0 scipy==1.7.1 absl-py einops
pip install pytorch_lightning==2.0.4 fair-esm mdtraj==1.9.9 wandb

**Comment by xtzhang0216** [comment]
when using the pl_upgrades branch, there will be an openfold err
`TypeError: __init__() missing 2 required positional arguments: 'opm_first' and 'fuse_projection_weights'`
which has been discussed in https://github.com/bjing2016/alphaflow/issues/2. It seems that we can only download openfold from 
`pip install 'openfold @ git+https://github.com/aqlaboratory/openfold.git@103d037'`

**Comment by xtzhang0216** [comment]
Has anyone had this problem? I hope someone can help me fix this problem. Thanks in advance

**Comment by agheeraert** [comment]
When running the commands that @MKCarter suggests, due to new version incompatibilites I got the infamous pytorch error
`libcusparse.so.12: undefined symbol: __nvJitLinkComplete_12_4, version libnvJitLink.so.12`
When updating the pytorch installation line to
`pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu126`
the installation worked just fine. However, then running the `predict.py` still raised a few errors since in pytorch 2.6 the weights_only default argument from torch.load was changed from `True` to `False`. After manually changing two incriminated lines in `predict.py` the prediction procedure appears to work (I'll edit when the first batch of results is complete).
